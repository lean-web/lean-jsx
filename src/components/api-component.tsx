import { registerAPIComponent } from "./component-registry";
import type { Request } from "express";

interface APIComponentConfig<Props extends Record<string, unknown>> {
  /**
   * The ID used for the API-facing component. It must be unique.
   */
  id: string;
  /**
   * A function to parse the Express request into the component parameters.
   * @param req - The Express request
   * @returns an object containing the component's parameters. Each object attribute is mapped as a component property.
   */
  requestHandler: (req: Request) => Props | Promise<Props>;
  /*
   * The server's response "cache" header (e.g. "cache: "public, max-age=30"")
   */
  cache?: string;
}

/**
 * Utility to register an API-facing component.
 * @template Props Type for the component properties
 * @template Component The component to render
 * @returns A decorated version of the passed JSX component that can be used in other components.
 */
export class APICBuilder<
  Props extends Record<string, unknown>,
  Component extends SXL.Element,
> {
  private id: string;
  private requestHandler: (req: Request) => Props | Promise<Props>;
  private cache = "no-cache";

  constructor(id: string, handler?: (req: Request) => Props | Promise<Props>) {
    this.id = id;
    this.requestHandler =
      handler ?? ((_req: Request) => this.getDefaultProps());
  }

  private getDefaultProps(): Props | Promise<Props> {
    return {} as Props;
  }

  /**
   * Set cache configuration for requesting this API Component
   * @param cache A string value for the response header "cache". Defaults to "no-cache"
   */
  withCache(cache: string): APICBuilder<Props, Component> {
    this.cache = cache;
    return this;
  }

  /**
   * Set the component to render and return it for use.
   * @param component A component to render
   * @returns A component configured as an API Component
   */
  render(
    component: (props: Props) => Component,
  ): (props: Props) => SXL.StaticElement {
    // if (!this.requestHandler) {
    //   this.requestHandler = (_req) => this.getDefaultProps();
    // }
    registerAPIComponent<NonNullable<Props>>({
      Api: component,
      contentId: this.id,
      requestHandler: this.requestHandler,
      cache: this.cache,
    });

    return (props: Props) => (
      <div ref={this.id} data-lean-api-component>
        {component(props)}
      </div>
    );
  }
}

/**
 * Utility to register an API-facing component.
 * @template Props Type for the component properties
 * @template Component The component to render
 * @param config - The API configuration for the endpoint exposed by the component
 * @param component - A component function
 * @returns A decorated version of the passed JSX component that can be used in other components.
 */
export function APIC<
  Props extends Record<string, unknown>,
  Component extends SXL.Element,
>(
  config: APIComponentConfig<NonNullable<Props>>,
  component: (props: Props) => Component,
) {
  registerAPIComponent<NonNullable<Props>>({
    Api: component,
    contentId: config.id,
    requestHandler: config.requestHandler,
    cache: config.cache,
  });

  return (props: Props) => (
    <div ref={config.id} data-lean-api-component>
      {component(props)}
    </div>
  );
}
