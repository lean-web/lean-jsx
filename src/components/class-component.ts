export abstract class Component<Props extends object = object>
  implements SXL.ClassComponent<Props & SXL.Props>
{
  props: Props & SXL.Props;
  constructor(props: Props & SXL.Props) {
    this.props = props;
  }
  abstract render(): SXL.StaticElement | SXL.AsyncElement;
}
