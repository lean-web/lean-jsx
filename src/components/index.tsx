/* eslint-disable @typescript-eslint/no-namespace */
import { isAsyncGen } from "@/jsx/html/jsx-utils";
import { SXLGlobalContext } from "lean-jsx-types/lib/context";
import { registerDynamicController } from "./component-registry";
import { Request } from "express";
export { withClientData } from "lean-web-utils/lib/server";
export { APIComponent } from "./api-component";

/**
 * Convert the contents of the global context into a valid URL.
 * @param url - the base URL to append the global context.
 * @param globalContext - the global context
 * @returns
 */
export function toQueryString<G extends SXLGlobalContext = SXLGlobalContext>(
  baseUrl: string,
  globalContext?: G,
): string {
  if (!globalContext) {
    return baseUrl;
  }
  if (baseUrl.includes("?")) {
    throw new Error("The base URL should not contain any query parameters");
  }
  return (
    baseUrl +
    "?" +
    Object.entries(globalContext)
      .filter(([key, value]) => !!key && !!value)
      .map(([key, value]) => `${key}=${value}`)
      .join("&")
  );
}

/**
 * The properties passed to {@link Lazy}.
 */
interface ComponentArgs extends SXL.Props {
  /**
   * A JSX component to render as a placeholder while the <Lazy>
   * main children is resolved.
   */
  loading: SXL.Element;
}

function isPromise(jsx: SXL.Element | undefined): jsx is SXL.AsyncElement {
  return !!jsx && "then" in jsx;
}

export abstract class Component<Props extends object = object>
  implements SXL.ClassComponent<Props & SXL.Props>
{
  props: Props & SXL.Props;
  constructor(props: Props & SXL.Props) {
    this.props = props;
  }
  abstract render(): SXL.StaticElement | SXL.AsyncElement;
}

/**
 * A utility JSX component that allows async JSX components to display
 * loading state as a placeholder.
 */
export class Lazy extends Component<ComponentArgs> {
  onLoading(): SXL.StaticElement {
    const props = this.props;
    if (isPromise(props.loading)) {
      return <></>;
    }
    if (isAsyncGen(props.loading)) {
      return <></>;
    }
    return props.loading;
  }

  async render(): SXL.AsyncElement {
    const props = this.props;
    if (!props.children) {
      throw new Error(
        "There is no child in Lazy to render. This is probably a mistake. If not, please file a bug.",
      );
    }
    const allResolved = await Promise.all(props.children);
    return <>{allResolved}</>;
  }
}

interface PendingResolve {
  isPending: true;
  isResolved: false;
  isError: false;
  value: null;
}

interface ResolvedPromise<T> {
  isPending: false;
  isResolved: true;
  isError: false;
  value: T;
}

/**
 * A wrapper around {@link Promise}, which exposes its state:
 * - isPending: The promise has not been resolved yet
 * - isResolved: The promise is fulfilled
 * - isError: The promise completed with an error
 * - value: The underlying value for a fulfilled promise.
 */
export type TrackedPromise<T> = PendingResolve | ResolvedPromise<T>;

/**
 * Provides methods to render a dynamically-loaded component
 * (one that doesn't block the main document loading).
 *
 * This object is meant to be created by the factory method {@link GetDynamicComponent},
 * as it relies on a very specific implementation.
 */
export interface DynamicController<
  GContext extends SXLGlobalContext = SXLGlobalContext,
  Props extends SXL.Props<object, GContext> = SXL.Props<object, GContext>,
> {
  contentId: string;
  /**
   * Render the component's loading state.
   * @param props - the component's properties {@link SXL.Props}
   * @returns - a JSX component
   */
  Render?: (props: Props) => SXL.Element;
  /**
   * Renders the component's loaded state.
   * @param props - the component's properties {@link SXL.Props}
   * @returns - a JSX component
   */
  Api: (
    props: Props,
  ) => SXL.AsyncElement | SXL.StaticElement | SXL.AsyncGenElement;

  /**
   * The parameters associated to this component
   */
  requestHandler?: (request: Request) => Record<string, unknown>;

  cache?: string;
}

/**
 * Creates a {@link DynamicController} wrapper for a dynamic component.
 *
 * Dynamic Components don't block the DOM, and they are similar to traditional React components.
 * The difference is that no state is tracked in the client, and the whole component is replaced
 * when the promise returned by a fetcher is fulfilled.
 *
 * @param contentId - A unique name for the component (e.g. "product-list")
 * @param fetcher - A function that returns a Promise.
 * @param render - A function that will receive the promise returned by the fetcher,
 *  and can return different JSX content depending on the promise state.
 * @returns A {@link DynamicController}
 */
export function GetDynamicComponent<
  T,
  GContext extends SXLGlobalContext = SXLGlobalContext,
  Props extends SXL.Props<object, GContext> = SXL.Props<object, GContext>,
>(
  contentId: string,
  fetcher: (props: Props) => Promise<T>,
  render: (
    data: TrackedPromise<T>,
    props: Props,
  ) => SXL.StaticElement | SXL.AsyncElement,
  queryParams?: (request: Request) => Record<string, string | number | boolean>,
): DynamicController<GContext, Props> {
  return registerDynamicController({
    Render: (props: Props) => (
      <dynamic-component
        data-id={toQueryString(contentId, props.globalContext)}
      >
        {render(
          {
            isPending: true,
            isError: false,
            isResolved: false,
            value: null,
          },
          props,
        )}
      </dynamic-component>
    ),
    Api: async (props: Props) => {
      const data = await fetcher(props);
      return render(
        {
          isPending: false,
          isError: false,
          isResolved: true,
          value: data,
        },
        props,
      );
    },
    contentId,
    requestHandler: queryParams,
  });
}

/**
 * A class decorator to auto register a dynamic component to the LeanJSX middleware
 * @param constructor: The class component extending {@link DynamicComponent}
 * @returns a decorated class
 */
export function Register<
  T extends DCConstructor<unknown, SXLGlobalContext, SXL.Props>,
>(constructor: T, _arg?: unknown): T {
  const ddc = new constructor({} as SXL.Props);
  registerDynamicController({
    Render: () => (
      <dynamic-component
        data-id={toQueryString(
          constructor.componentID,
          ddc.props.globalContext,
        )}
      >
        {ddc.dynamicRender(
          {
            isPending: true,
            isError: false,
            isResolved: false,
            value: null,
          },
          ddc.props,
        )}
      </dynamic-component>
    ),
    Api: async (props: SXL.Props) => {
      const ddc = new constructor(props);
      const data = await ddc.fetcher(props);
      return ddc.dynamicRender(
        {
          isPending: false,
          isError: false,
          isResolved: true,
          value: data,
        },
        props,
      );
    },
    contentId: ddc.componentID,
    requestHandler: ddc.queryParams.bind(ddc),
  });

  constructor.componentID = ddc.componentID;
  return constructor;
}

interface DCConstructor<
  T,
  GContext extends SXLGlobalContext = SXLGlobalContext,
  Props extends SXL.Props<object, GContext> = SXL.Props<object, GContext>,
> {
  new (props: Props): DynamicComponent<T, GContext, Props>;
  componentID: string;
}

/**
 * The base class for dynamic components.
 */
export abstract class DynamicComponent<
  T,
  GContext extends SXLGlobalContext = SXLGlobalContext,
  Props extends SXL.Props<object, GContext> = SXL.Props<object, GContext>,
> implements SXL.ClassComponent<SXL.Props<object, GContext>>
{
  props: Props;
  constructor(props: Props) {
    this.props = props;
  }

  /**
   * A string identifier for this component. It must be unique within a single application.
   */
  abstract componentID: string;
  declare static componentID: string;

  /**
   * A function to define the asynchronous data that this component depends on.
   * @param props - the props passed to the component, along with the global context, which will
   * contain the query parameters defined by the component.
   */
  abstract fetcher(props: Props): Promise<T>;

  /**
   * The main render function.
   * @param data - A {@link TrackedPromise} containing the data returned by the fetcher.
   * @param props - the props passed to the component, along with the global context, which will
   * contain the query parameters defined by the component.
   */
  abstract dynamicRender(
    data: TrackedPromise<T>,
    props: Props,
  ): SXL.StaticElement | SXL.AsyncElement;

  /**
   * A function to parse query parameters from the request. These query params are specific for this component.
   * @param _request
   * @returns An object containing the parsed query parameters.
   */
  queryParams(_request: Request) {
    return {};
  }

  /**
   * The actual rendering function. This calls dynamicRender under the hood.
   * @param props the props passed to the component
   * @returns the components JSX content
   */
  render() {
    const props = this.props;
    return (
      <dynamic-component
        data-id={toQueryString(this.componentID, props.globalContext)}
      >
        {this.dynamicRender(
          {
            isPending: true,
            isError: false,
            isResolved: false,
            value: null,
          },
          props,
        )}
      </dynamic-component>
    );
  }
}

declare global {
  namespace JSX {
    // we override the IntinsicElements interface to include the web component dynamic-component
    interface IntrinsicElements extends SXL.IntrinsicElements {
      "dynamic-component": Partial<HTMLElement>;
    }
  }
}
