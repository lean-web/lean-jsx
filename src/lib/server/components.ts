export * from "@/components/index";
import type { IWebActions } from "lean-jsx-types/events";

/**
 * Utility that allows component developers to explicitely pass data to the browser
 * for a given event handler (e.g. onclick) in a JSX element.
 *
 * @param data the data to pass to the browser
 * @param handler - a function that receives two parameters (on the client side):
 * - The Event object from the listener.
 * - An object containing API-components actions
 * - A context object with access to the server-passed data.
 * @returns a web action handler, used internall by LeanJSX
 *  during render.
 */
export function withClientData<
    Ev extends Event,
    Data extends Record<string, unknown>,
>(
    data: Data,
    handler: (
        this: Element | GlobalEventHandlers | null,
        ev: Ev,
        actions: IWebActions,
        data: Data,
    ) => unknown,
): SXL.WebHandler<Ev, Data> {
    return {
        handler: handler,
        data,
    };
}
