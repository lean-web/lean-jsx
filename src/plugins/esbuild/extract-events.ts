import ts from "typescript";
import crypto from "crypto";
import { HANDLERS_NAMESPACE } from "../constants";

/**
 * Replace the handler function's body with a reference to a global reference to the handler
 *
 * @param handler - The expression for the handler function
 * @param ident - The identifier with the reference to the function definition
 * @returns - A new handler function which references a global handler definition.
 */
function replaceHandlerContent(handler: ts.Expression, ident: ts.Expression) {
  if (!(ts.isFunctionExpression(handler) || ts.isArrowFunction(handler))) {
    return handler;
  }
  const newBody = ts.factory.createBlock([
    ts.factory.createReturnStatement(
      ts.factory.createCallExpression(
        ident,
        undefined,
        handler.parameters.map((p) =>
          ts.factory.createIdentifier(p.name.getText()),
        ),
      ),
    ),
  ]);

  if (ts.isFunctionExpression(handler)) {
    return ts.factory.updateFunctionExpression(
      handler,
      handler.modifiers,
      handler.asteriskToken,
      handler.name,
      handler.typeParameters,
      handler.parameters,
      handler.type,
      newBody,
    );
  }

  if (ts.isArrowFunction(handler)) {
    return ts.factory.updateArrowFunction(
      handler,
      handler.modifiers,
      handler.typeParameters,
      handler.parameters,
      handler.type,
      handler.equalsGreaterThanToken,
      newBody,
    );
  }

  return handler;
}

/**
 * Create a unique name for a function handler.
 * @param fn - the text for the function definition
 * @returns - a string with a unique function name
 */
function generateFunctionName(fn: string) {
  const hash = crypto.createHash("md5").update(fn).digest("hex");
  return `fn${hash}`;
}

interface EventExtractorOptions {
  skipShortHandlers: boolean;
}

export class EventExtractor {
  header = `import { withClientData } from "lean-web-utils/lib/server";\ntype Arg = Parameters<typeof withClientData>[1]`;

  handlers: Record<string, string[]> = {};
  existingHandlers: Set<string> = new Set();
  skipShortHandlers: boolean;

  constructor(options: EventExtractorOptions = { skipShortHandlers: false }) {
    this.skipShortHandlers = options.skipShortHandlers;
  }

  fileTransformer(source: string, filePath = "file.tsx") {
    const printer = ts.createPrinter({
      newLine: ts.NewLineKind.LineFeed,
    });

    const { existingHandlers, handlers, skipShortHandlers } = this;

    const sourceFile = ts.createSourceFile(
      filePath,
      source,
      ts.ScriptTarget.Latest,
      true,
    );

    const transformer =
      (context: ts.TransformationContext) =>
      (rootNode: ts.SourceFile): ts.SourceFile => {
        function visit<T extends ts.Node>(node: T): T {
          if (
            ts.isJsxSelfClosingElement(node) ||
            ts.isJsxOpeningElement(node)
          ) {
            const newAttributes = node.attributes.properties.map(
              (prop: ts.JsxAttributeLike) => {
                if (
                  ts.isJsxAttribute(prop) &&
                  !!prop.name &&
                  ts.isIdentifier(prop.name) &&
                  !!prop.name.escapedText.toString().match(/^on/) &&
                  !!prop.initializer &&
                  ts.isJsxExpression(prop.initializer) &&
                  !!prop.initializer.expression
                ) {
                  if (
                    ts.isCallExpression(prop.initializer.expression) &&
                    ts.isIdentifier(prop.initializer.expression.expression) &&
                    prop.initializer.expression.expression.escapedText.toString() ===
                      "withClientData"
                  ) {
                    const fnText =
                      prop.initializer.expression.arguments[1].getText();
                    const fn = generateFunctionName(fnText);

                    // Create a property chain identifier: e.g.: window.sxl.fn1
                    const access = ts.factory.createPropertyAccessExpression(
                      ts.factory.createPropertyAccessExpression(
                        ts.factory.createIdentifier("window"),
                        HANDLERS_NAMESPACE,
                      ),
                      fn,
                    );

                    // if the generated function name is longer than the handler itself,
                    // replacing it would not improve performance, so we skip those listeners:
                    if (
                      skipShortHandlers &&
                      printer.printNode(
                        ts.EmitHint.Unspecified,
                        access,
                        sourceFile,
                      ).length < fn.length
                    ) {
                      return prop;
                    }

                    if (!existingHandlers.has(fn)) {
                      // capture the handlers code, and append it to the "sxl" global namespace.
                      if (!handlers[filePath]) {
                        handlers[filePath] = [];
                      }
                      handlers[filePath].push(
                        `export const ${fn} = ${fnText};\n`,
                      );
                      //   eventHandlersBundle += `window.${NAMESPACE}.${fn} = ${fnText};\n`;
                      existingHandlers.add(fn);
                    }

                    // Replace the body contents:
                    const nfn = replaceHandlerContent(
                      prop.initializer.expression.arguments[1],
                      access,
                    );
                    const callExpression = ts.factory.updateCallExpression(
                      prop.initializer.expression,
                      prop.initializer.expression.expression,
                      prop.initializer.expression.typeArguments,
                      [prop.initializer.expression.arguments[0], nfn],
                    ) as T & ts.CallExpression;

                    return ts.factory.updateJsxAttribute(
                      prop,
                      prop.name,
                      ts.factory.updateJsxExpression(
                        prop.initializer,
                        callExpression,
                      ),
                    );
                  } else if (
                    ts.isArrowFunction(prop.initializer.expression) ||
                    ts.isFunctionDeclaration(prop.initializer.expression)
                  ) {
                    const fnText = prop.initializer.expression.getText();
                    const fn = generateFunctionName(fnText);

                    // Create a property chain identifier: e.g.: window.sxl.fn1
                    const access = ts.factory.createPropertyAccessExpression(
                      ts.factory.createPropertyAccessExpression(
                        ts.factory.createIdentifier("window"),
                        HANDLERS_NAMESPACE,
                      ),
                      fn,
                    );

                    // if the generated function name is longer than the handler itself,
                    // replacing it would not improve performance, so we skip those listeners:
                    if (
                      skipShortHandlers &&
                      printer.printNode(
                        ts.EmitHint.Unspecified,
                        access,
                        sourceFile,
                      ).length < fn.length
                    ) {
                      return prop;
                    }

                    if (!existingHandlers.has(fn)) {
                      // capture the handlers code, and append it to the "sxl" global namespace.
                      if (!handlers[filePath]) {
                        handlers[filePath] = [];
                      }
                      handlers[filePath].push(
                        `export const ${fn} = ${fnText};\n`,
                      );
                      //   eventHandlersBundle += `window.${NAMESPACE}.${fn} = ${fnText};\n`;
                      existingHandlers.add(fn);
                    }

                    // Replace the body contents:
                    const expression = replaceHandlerContent(
                      prop.initializer.expression,
                      access,
                    );

                    return ts.factory.updateJsxAttribute(
                      prop,
                      prop.name,
                      ts.factory.updateJsxExpression(
                        prop.initializer,
                        expression,
                      ),
                    );
                  } else {
                    return prop;
                  }
                }
                return prop;
              },
            );

            if (ts.isJsxSelfClosingElement(node)) {
              return ts.factory.updateJsxSelfClosingElement(
                node,
                node.tagName,
                node.typeArguments,
                ts.factory.updateJsxAttributes(node.attributes, newAttributes),
              ) as T & ts.JsxSelfClosingElement;
            }

            if (ts.isJsxOpeningElement(node)) {
              return ts.factory.updateJsxOpeningElement(
                node,
                node.tagName,
                node.typeArguments,
                ts.factory.updateJsxAttributes(node.attributes, newAttributes),
              ) as T & ts.JsxOpeningElement;
            }
          }

          return ts.visitEachChild(node, visit, context);
        }
        return ts.visitNode(rootNode, visit) as ts.SourceFile;
      };

    const result = ts.transform(sourceFile, [transformer]);
    return printer.printFile(result.transformed[0]);
  }

  //   getActionsFile: () => eventHandlersBundle,
  getActionsPerFile() {
    return Object.entries(this.handlers).map(([fpath, handlrs]) => [
      fpath,
      `
          ${this.header}
          // ${fpath}
          ${handlrs.join("\n")}
      `.replace(/\n[\s\t]+/g, "\n"),
    ]);
  }

  clear() {
    this.handlers = {};
    this.existingHandlers = new Set();
  }
}
