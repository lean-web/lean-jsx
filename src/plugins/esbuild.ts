import { type Plugin } from "esbuild";
console.log("Import fs");
import fs from "fs";
import path from "path";
import { EventExtractor } from "./esbuild/extract-events";
import glob from "fast-glob";
import { HANDLERS_NAMESPACE } from "./constants";

/**
 * A plugin which finds all event handlers used with `withClientData` and bundles them in a JS accessible by the browser.
 *
 * This is necessary because JSX components that render elements with event handlers in a loop
 * will pass their handler code as it is to the browser, increasing the response size by passing a copy of the
 * handler for each element in the loop. This could be very inefficient for large event handlers.
 *
 * This pre-processing help us create a script with a single instance of each event handler,
 * and the script injected by element in a loop would just reference the event handler definition.
 */
const extractEventHandlersPlugin: Plugin = {
  name: "extract-event-handlers",
  setup(build) {
    const extractor = new EventExtractor();

    build.onLoad({ filter: /\.tsx?$/ }, async (args) => {
      const source = await fs.promises.readFile(args.path, "utf8");

      const transformedSource = extractor.fileTransformer(source, args.path);

      return {
        contents: transformedSource,
        loader: "tsx",
      };
    });

    build.onEnd(async (result) => {
      if (result.errors.length > 0) {
        return;
      }
      if (!fs.existsSync(".lean/handlers/")) {
        fs.mkdirSync(".lean/handlers/", { recursive: true });
      }
      const filesAndActions = extractor.getActionsPerFile();
      if (filesAndActions.length === 0) {
        return;
      }

      const cachePath = path.join(process.cwd(), ".lean");
      const files = fs.readdirSync(cachePath);

      for (const f of files) {
        if (f.includes("action-handlers")) {
          fs.rmSync(path.join(cachePath, f));
        }
      }

      await Promise.all(
        filesAndActions
          .map(([fpath, content]) => {
            return [
              path.resolve(
                ".lean/handlers/",
                path.basename(fpath).replace(".tsx", ".ts"),
              ),
              content,
            ];
          })
          .map(([fpath, fcontents]) => fs.promises.writeFile(fpath, fcontents)),
      );

      const fileContents = await Promise.all(
        glob
          .sync(".lean/handlers/**/*.ts")
          .map(async (fpath) => fs.promises.readFile(fpath, "utf8")),
      );

      await build.esbuild.build({
        stdin: { contents: fileContents.join("\n"), loader: "tsx" },
        // outfile: ".lean/action-handlers.js",
        outdir: ".lean",
        entryNames: `[dir]/action-handlers-[hash]`,
        platform: "browser", // Set to 'browser' for IIFE format
        target: ["es2020"], // Target modern browsers or environments
        bundle: true, // Bundle all dependencies into one file
        format: "iife", // Output format as IIFE
        globalName: HANDLERS_NAMESPACE, // Global variable for the IIFE
        external: ["esbuild", "lean-web-utils/lib/server"], // Exclude esbuild from the bundle
      });

      extractor.clear();
    });
  },
};

export default extractEventHandlersPlugin;
