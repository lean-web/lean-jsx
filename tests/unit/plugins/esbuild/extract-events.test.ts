import { EventExtractor } from "@/plugins/esbuild/extract-events";
import { describe, expect, test } from "@jest/globals";
import { validateTypeScriptCode } from "./validate-ts";
import path from "path";
import fs from "fs";

function validateHandlersExist(source: string, extractor: EventExtractor) {
  return (
    Array.from(extractor.existingHandlers.values()).filter((handler) =>
      source.includes(handler),
    ).length === extractor.existingHandlers.size
  );
}

const tsConfig = path.join(process.cwd(), "./packages/core/tsconfig.json");

describe("extract-events.test", () => {
  let extractor: EventExtractor;
  beforeAll(() => {
    extractor = new EventExtractor();
  });

  afterEach(() => {
    extractor.clear();
  });

  test("ts event extractor find a single action", () => {
    const source = `
      export function MyComponent() {
        return <button onclick={(ev) => console.log(ev)}>Click</button> 
      }
        `;

    expect(validateTypeScriptCode(source, tsConfig)).toBeFalsy();
    const transformed = extractor.fileTransformer(source);

    expect(transformed).toMatchInlineSnapshot(`
      "export function MyComponent() {
          return <button onclick={(ev) => { return window.sxl_user.fn14533fb3b06a348bf24a9f456f6b9275(ev); }}>Click</button>;
      }
      "
    `);

    expect(
      validateTypeScriptCode(
        `declare global {
          interface Window { sxl_user: any; }
      }\n${transformed}`,
        tsConfig,
      ),
    ).toBeFalsy();

    expect(extractor.getActionsPerFile()).toMatchInlineSnapshot(`
      [
        [
          "file.tsx",
          "
      import { withClientData } from "lean-web-utils/server";
      type Arg = Parameters<typeof withClientData>[1]
      // file.tsx
      export const fn14533fb3b06a348bf24a9f456f6b9275 = (ev) => console.log(ev);
      ",
        ],
      ]
    `);
  });

  test("ts event extractor find a webAction action", () => {
    const source = `
    import {  withClientData } from "@/components";
    function MyComponent() {
        const user = { firstName: "John" };
        return (
            <button
                onclick={withClientData(user, (ev, webContext) => {
                    alert(\`Hello \${webContext?.data.firstName}\`);
                })}
            >
                Click to greet user
            </button>
        );
    }
  `;

    expect(validateTypeScriptCode(source, tsConfig)).toBeFalsy();
    const transformed = extractor.fileTransformer(source);

    expect(validateHandlersExist(transformed, extractor)).toBeTruthy();

    expect(transformed).toMatchInlineSnapshot(`
      "import { withClientData } from "@/components";
      function MyComponent() {
          const user = { firstName: "John" };
          return (<button onclick={withClientData(user, (ev, webContext) => { return window.sxl_user.fn6dff4efb11ab2eedbca9f29835365cfb(ev, webContext); })}>
                      Click to greet user
                  </button>);
      }
      "
    `);

    expect(
      validateTypeScriptCode(
        `declare global {
            interface Window { sxl_user: any; }
        }\n${transformed}`,
        tsConfig,
      ),
    ).toBeFalsy();

    const actions = extractor
      .getActionsPerFile()
      .map(([_, contents]) => contents)
      .join(";");
    expect(actions).toMatchInlineSnapshot(`
      "
      import { withClientData } from "lean-web-utils/server";
      type Arg = Parameters<typeof withClientData>[1]
      // file.tsx
      export const fn6dff4efb11ab2eedbca9f29835365cfb = (ev, webContext) => {
      alert(\`Hello \${webContext?.data.firstName}\`);
      };
      "
    `);

    expect(
      validateTypeScriptCode(
        `declare global {
        interface Window { sxl_user: any; }
    }\n${actions}`,
        tsConfig,
      ),
    ).toBeFalsy();
  });

  test("ts event extractor - large file", () => {
    const source = fs.readFileSync(
      path.resolve(
        "packages/core/tests/unit/plugins/esbuild/test-data/large-component.tsx",
      ),
      "utf8",
    );

    expect(
      validateTypeScriptCode(
        source,
        tsConfig,
        "packages/core/tests/unit/plugins/esbuild/test-data/large-component.tsx",
        [
          path.resolve(
            "packages/core/tests/unit/plugins/esbuild/test-data/extra-components.tsx",
          ),
        ],
      ),
    ).toBeFalsy();
    const transformed = extractor.fileTransformer(source);

    expect(transformed).toMatchSnapshot();

    expect(validateHandlersExist(transformed, extractor)).toBeTruthy();

    expect(
      validateTypeScriptCode(
        `declare global {
            interface Window { sxl_user: any; }
        }\n${transformed}`,
        tsConfig,
      ),
    ).toBeFalsy();

    const actions = extractor
      .getActionsPerFile()
      .map(([_, contents]) => contents)
      .join(";");
    expect(actions).toMatchSnapshot();

    expect(
      validateTypeScriptCode(
        `declare global {
        interface Window { sxl_user: any; }
    }\n${actions}`,
        tsConfig,
      ),
    ).toBeFalsy();
  });

  test("ts event extractor finds more event listeners besides onclick", () => {
    // notice onchange and onload have the exact same function,
    // so we only generate one handler for both:
    const source = `
      export function MyComponent() {
        return <button onclick={(c1) => c1} onchange={(c) => c} onload={(c) => c}>Click</button> 
      }
        `;

    expect(validateHandlersExist(source, extractor)).toBeTruthy();
    const transformed = extractor.fileTransformer(source);
    expect(validateHandlersExist(transformed, extractor)).toBeTruthy();

    expect(transformed).toMatchInlineSnapshot(`
      "export function MyComponent() {
          return <button onclick={(c1) => { return window.sxl_user.fnc811db883db9743ff5764cd85fb9d030(c1); }} onchange={(c) => { return window.sxl_user.fn65934597b0f8eb9debfc0762c97da7a6(c); }} onload={(c) => { return window.sxl_user.fn65934597b0f8eb9debfc0762c97da7a6(c); }}>Click</button>;
      }
      "
    `);

    expect(extractor.getActionsPerFile()).toMatchInlineSnapshot(`
      [
        [
          "file.tsx",
          "
      import { withClientData } from "lean-web-utils/server";
      type Arg = Parameters<typeof withClientData>[1]
      // file.tsx
      export const fnc811db883db9743ff5764cd85fb9d030 = (c1) => c1;
      export const fn65934597b0f8eb9debfc0762c97da7a6 = (c) => c;
      ",
        ],
      ]
    `);
  });

  test("we can skip processing short handlers", () => {
    extractor = new EventExtractor({ skipShortHandlers: true });
    const source = `
      export function MyComponent() {
        return <button onclick={(c1) => console.log('this is a long enough event handler')} onchange={(c) => c} onload={(c) => c}>Click</button> 
      }
        `;

    expect(validateHandlersExist(source, extractor)).toBeTruthy();
    const transformed = extractor.fileTransformer(source);
    expect(validateHandlersExist(transformed, extractor)).toBeTruthy();

    expect(transformed).toMatchInlineSnapshot(`
      "export function MyComponent() {
          return <button onclick={(c1) => { return window.sxl_user.fn8a6ef6b2b95f5134616d30c92a027aa6(c1); }} onchange={(c) => { return window.sxl_user.fn65934597b0f8eb9debfc0762c97da7a6(c); }} onload={(c) => { return window.sxl_user.fn65934597b0f8eb9debfc0762c97da7a6(c); }}>Click</button>;
      }
      "
    `);

    expect(extractor.getActionsPerFile()).toMatchInlineSnapshot(`
      [
        [
          "file.tsx",
          "
      import { withClientData } from "lean-web-utils/server";
      type Arg = Parameters<typeof withClientData>[1]
      // file.tsx
      export const fn8a6ef6b2b95f5134616d30c92a027aa6 = (c1) => console.log('this is a long enough event handler');
      export const fn65934597b0f8eb9debfc0762c97da7a6 = (c) => c;
      ",
        ],
      ]
    `);
  });
});
