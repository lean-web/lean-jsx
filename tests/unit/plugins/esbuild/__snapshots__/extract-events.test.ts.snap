// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract-events.test ts event extractor - large file 1`] = `
"import { withClientData } from "@/components";
export function DeeplyNestedComponent() {
    return (<button onclick={withClientData({}, (ev, actions) => { return window.sxl_user.fn7377f93f0e35f7fe806e1e81120aeab5(ev, actions); })}>
      Click to greet user
    </button>);
}
/**
 * Contains a handler with the exact same code as DeeplyNestedComponent
 * @returns
 */
export function DeeplyNestedComponent2() {
    return (<button onclick={withClientData({}, (ev, actions) => { return window.sxl_user.fn7377f93f0e35f7fe806e1e81120aeab5(ev, actions); })}>
      Click to greet user
    </button>);
}
export function NestedComponent() {
    return (<div>
      <DeeplyNestedComponent />
      <button onclick={withClientData({}, (ev, actions) => { return window.sxl_user.fn0722c48dbc0caab0969e3b183c574489(ev, actions); })}>
        Click to greet user
      </button>
      <DeeplyNestedComponent2 />
    </div>);
}
function LoopedComponent() {
    return (<>
      {new Array(5).fill(true).map((_, index) => (<button onclick={withClientData({ index }, (ev, _, data) => { return window.sxl_user.fnbdf4b7850b98807cc4e6a97b676403dc(ev, _, data); })}>
          Click {index}
        </button>))}
    </>);
}
export function MyComponent() {
    const user = { firstName: "John" };
    return (<div>
      <NestedComponent />
      <LoopedComponent />
      <button onclick={withClientData(user, (ev, actions, data) => { return window.sxl_user.fn6baf69addf06de37f253545936c58b1a(ev, actions, data); })}>
        Click to greet user
      </button>
    </div>);
}
"
`;

exports[`extract-events.test ts event extractor - large file 2`] = `
"
import { withClientData } from "lean-web-utils/server";
type Arg = Parameters<typeof withClientData>[1]
// file.tsx
export const fn7377f93f0e35f7fe806e1e81120aeab5 = (ev, actions) => {
void actions.refetchAPIC("my-element2", {});
};
export const fn0722c48dbc0caab0969e3b183c574489 = (ev, actions) => {
void actions.refetchAPIC("my-element", {});
};
export const fnbdf4b7850b98807cc4e6a97b676403dc = (ev, _, data) =>
console.log(data.index);
export const fn6baf69addf06de37f253545936c58b1a = (ev, actions, data) => {
alert(\`Hello \${data.firstName}\`);
};
"
`;
